modules/client/src/main/java/org/apache/ignite/internal/client/sql/ClientAsyncResultSet.java

```diff
@@ -0,0 +1,196 @@
+import org.apache.ignite.internal.client.ClientChannel;
+import org.apache.ignite.internal.client.proto.ClientMessageUnpacker;
+import org.apache.ignite.internal.client.proto.ClientOp;
+import org.apache.ignite.sql.NoRowSetExpectedException;
+import org.apache.ignite.sql.ResultSetMetadata;
+import org.apache.ignite.sql.SqlRow;
+import org.apache.ignite.sql.async.AsyncResultSet;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Client async result set.
+ */
+class ClientAsyncResultSet implements AsyncResultSet {
+    /** Channel. */
+    private final ClientChannel ch;
+
+    /** Resource id. */
+    private final Long resourceId;
+
+    /** Row set flag. */
+    private final boolean hasRowSet;
+
+    /** Applied flag. */
+    private final boolean wasApplied;
+
+    /** Affected rows. */
+    private final long affectedRows;
+
+    /** Metadata. */
+    private final ResultSetMetadata metadata;
+
+    /** Rows. */
+    private volatile List<SqlRow> rows;
+
+    /** More pages flag. */
+    private volatile boolean hasMorePages;
+
+    /** Closed flag. */
+    private volatile boolean closed;
+
+    /**
+     * Constructor.
+     *
+     * @param ch Channel.
+     * @param in Unpacker.
+     */
+    public ClientAsyncResultSet(ClientChannel ch, ClientMessageUnpacker in) {
+        this.ch = ch;
+
+        resourceId = in.tryUnpackNil() ? null : in.unpackLong();
+        hasRowSet = in.unpackBoolean();
+        hasMorePages = in.unpackBoolean();
+        wasApplied = in.unpackBoolean();
+        affectedRows = in.unpackLong();
+
+        metadata = new ClientResultSetMetadata(in);
+
+        if (hasRowSet) {
+            readRows(in);
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public @Nullable ResultSetMetadata metadata() {
+        return metadata;
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public boolean hasRowSet() {
+        return hasRowSet;
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public long affectedRows() {
+        return affectedRows;
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public boolean wasApplied() {
+        return wasApplied;
+    }
+
+    /** {@inheritDoc} */
+    @SuppressWarnings("AssignmentOrReturnOfFieldWithMutableType")
+    @Override
+    public Iterable<SqlRow> currentPage() {
+        requireResultSet();
+
+        return rows;
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public int currentPageSize() {
+        requireResultSet();
+
+        return rows.size();
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public CompletionStage<? extends AsyncResultSet> fetchNextPage() {
+        requireResultSet();
+
+        if (closed) {
+            return CompletableFuture.failedFuture(new IgniteClientException("Cursor is closed."));
+        }
+
+        if (!hasMorePages || resourceId == null) {
+            return CompletableFuture.failedFuture(new IgniteClientException("No more pages."));
+        }
```

Comment:
The check does not match the one in `hasMorePages()`, which may result in user calling `hasMorePages()`, getting `true`, trying fetch next page and getting an exception as a result (if `resourceId == null`). I know that maybe this is not possible scenario, but anyway, lets use `hasMorePages()` call here, and move `resourceId == null` check to this method.

Reply:
Agree, fixed.